### PR TITLE
Add support for expanding local volumes

### LINKED ISSUE #101340: Allow Node Only expansion of Local Volumes
There are use cases where users want to expand the local volume (may be managed by LVM or something similar), but even though they can edit underlying PV (and expand the underlying device) -  the reported size in pvc.Status keeps reporting older value because in-tree local volumes are non-expandable.

We could make this easier for users by - implementing `NodeExpandableVolumePlugin` interface for local volumes. 

This would allow following steps for local volume expansion:

1. User/admin edits the pvc - requesting new size.
2. Admin expands the underlying device.
2. Admin edits the PV and sets new size.
3. Kubelet will expand the filesystem and update `pvc.Status`

This would at least get the user from impossible territory  to possible.

cc @msau42 @cofyc 

/sig storage

### PR DESCRIPTION
**Pull request by gnufied**
Fixes https://github.com/kubernetes/kubernetes/issues/101340

This is another variant of https://github.com/kubernetes/kubernetes/pull/101415

for shared directory - we consider expansion operation as a NO-OP.

/sig storage

/assign @msau42 

```
I0615 13:46:04.773435 4149291 local.go:415] "expansion of directory based local volumes is NO-OP" local-volume-path="/mnt/shared"
I0615 13:46:04.773463 4149291 operation_generator.go:1777] MountVolume.NodeExpandVolume succeeded for volume "example-pv" (UniqueName: "kubernetes.io/local-volume/example-pv") pod "busybox" (UID: "6bdc0604-2e9d-4224-bf8b-70c876a1c295") 
I0615 13:46:04.773555 4149291 event.go:291] "Event occurred" object="default/busybox" kind="Pod" apiVersion="v1" type="Normal" reason="FileSystemResizeSuccessful" message="MountVolume.NodeExpandVolume succeeded for volume \"example-pv\" "
I0615 13:46:04.773577 4149291 event.go:291] "Event occurred" object="default/myclaim1" kind="PersistentVolumeClaim" apiVersion="v1" type="Normal" reason="FileSystemResizeSuccessful" message="MountVolume.NodeExpandVolume succeeded for volume \"example-pv\" "
```

```release-note
Allow node expansion of local volumes
```

### CODE REVIEW

**Comment by gnufied** [comment]
gentle nudge on this again @msau42

File: pkg/volume/local/local.go [hunk 1]
```diff
@@ -376,6 +377,48 @@ func (dm *deviceMounter) MountDevice(spec *volume.Spec, devicePath string, devic
...
+		return false, fmt.Errorf("plugin volume host does not implement KubeletVolumeHost interface")
+	}
+
+	fileType, err := kvh.GetHostUtil().GetFileType(localDevicePath)
+	if err != nil {
+		return false, err
+	}
+
+	switch fileType {
+	case hostutil.FileTypeBlockDev:
```
Comment (msau42) [inline pkg/volume/local/local.go:406]: can we add an e2e test for this? We have some tests that create a tmp device.

**Comment by dims** [comment]
@gnufied looks like this needs a new e2e test? ( https://github.com/kubernetes/kubernetes/pull/102886#discussion_r658377601 )

**Comment by gnufied** [comment]
@msau42 added e2e , PTAL.

File: pkg/volume/local/local.go [hunk 2]
```diff
@@ -376,6 +377,48 @@ func (dm *deviceMounter) MountDevice(spec *volume.Spec, devicePath string, devic
...
+		_, err = util.GenericResizeFS(plugin.host, plugin.GetPluginName(), localDevicePath, resizeOptions.DeviceMountPath)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	case hostutil.FileTypeDirectory:
```
Comment (msau42) [inline pkg/volume/local/local.go:412]: Is it possible to unit test or e2e test this condition as well?

Comment (gnufied) [reply pkg/volume/local/local.go:412]: done.

**Pull request merged**